quaternion_ball_joint: Bionic GCC errors on -Wmaybe-unused

CI Failure:
https://drake-jenkins.csail.mit.edu/view/Continuous%20Production/job/linux-bionic-gcc-bazel-continuous-release/40/

Excerpt:
```
18:06:44 ERROR: /home/ubuntu/workspace/linux-bionic-gcc-bazel-continuous-release/src/attic/multibody/joints/BUILD.bazel:23:1: Couldn't build file attic/multibody/joints/_objs/everything/quaternion_ball_joint.pic.o: C++ compilation of rule '//attic/multibody/joints:everything' failed (Exit 1)
18:06:44 In file included from bazel-out/k8-opt/bin/common/_virtual_includes/autodiff/drake/common/autodiff.h:31:0,
18:06:44                  from bazel-out/k8-opt/bin/common/_virtual_includes/drake_bool/drake/common/drake_bool_deprecated.h:12,
18:06:44                  from bazel-out/k8-opt/bin/common/_virtual_includes/drake_bool/drake/common/drake_bool.h:101,
18:06:44                  from bazel-out/k8-opt/bin/math/_virtual_includes/geometric_transform/drake/math/rotation_matrix.h:11,
18:06:44                  from bazel-out/k8-opt/bin/math/_virtual_includes/geometric_transform/drake/math/quaternion.h:13,
18:06:44                  from bazel-out/k8-opt/bin/attic/multibody/joints/_virtual_includes/everything/drake/multibody/joints/quaternion_ball_joint.h:10,
18:06:44                  from attic/multibody/joints/quaternion_ball_joint.cc:1:
18:06:44 bazel-out/k8-opt/bin/common/_virtual_includes/autodiff/drake/common/autodiffxd.h: In function 'drake::Matrix3<typename Derived::Scalar> drake::math::quat2rotmat(const Eigen::MatrixBase<Derived>&) [with Derived = Eigen::Matrix<Eigen::AutoDiffScalar<Eigen::Matrix<double, -1, 1> >, 4, 1, 0, 4, 1>]':
18:06:44 bazel-out/k8-opt/bin/common/_virtual_includes/autodiff/drake/common/autodiffxd.h:103:26: error: 'R' may be used uninitialized in this function [-Werror=maybe-uninitialized]
18:06:44      m_value = other.value();
```

Potentially the offending source:
https://github.com/RobotLocomotion/drake/blob/bbf54c97f33b5f9339518997042c6dee05dcd371/attic/multibody/joints/quaternion_ball_joint.h#L77

\cc @SeanCurtis-TRI @jamiesnape

--- a/math/rotation_matrix.h
+++ b/math/rotation_matrix.h
@@ -74,7 +74,7 @@ class RotationMatrix {
   // TODO(mitiguy) Although this method is fairly efficient, consider adding an
   // optional second argument if `quaternion` is known to be normalized apriori
   // or for some reason the calling site does not want `quaternion` normalized.
-  explicit RotationMatrix(const Eigen::Quaternion<T>& quaternion) : R_AB_() {
+  explicit RotationMatrix(const Eigen::Quaternion<T>& quaternion) {
     // Cost for various way to create a rotation matrix from a quaternion.
     // Eigen quaternion.toRotationMatrix() = 12 multiplies, 12 adds.
     // Drake  QuaternionToRotationMatrix() = 12 multiplies, 12 adds.
@@ -98,7 +98,7 @@ class RotationMatrix {
   // %RotationMatrix constructor (above) with that un-normalized quaternion.
   // TODO(mitiguy) Consider adding an optional second argument if `lambda` is
   // known to be normalized apriori or calling site does not want normalization.
-  explicit RotationMatrix(const Eigen::AngleAxis<T>& theta_lambda) : R_AB_() {
+  explicit RotationMatrix(const Eigen::AngleAxis<T>& theta_lambda) {
     const Vector3<T>& lambda = theta_lambda.axis();
     const T norm = lambda.norm();
     const T& theta = theta_lambda.angle();
@@ -136,7 +136,7 @@ class RotationMatrix {
   /// @li 3rd rotation R_AB: Frames D, C, B (collectively -- as if welded)
   /// rotate relative to frame A by a roll angle `y` about `Bz = Az`.
   /// Note: B and A are no longer aligned.
-  explicit RotationMatrix(const RollPitchYaw<T>& rpy) : R_AB_() {
+  explicit RotationMatrix(const RollPitchYaw<T>& rpy) {
     const T& r = rpy.roll_angle();
     const T& p = rpy.pitch_angle();
     const T& y = rpy.yaw_angle();
@@ -173,7 +173,7 @@ class RotationMatrix {
   /// columns of `R_AB` are `Bx`, `By`, `Bz`.
   static RotationMatrix<T> MakeFromOrthonormalColumns(
       const Vector3<T>& Bx, const Vector3<T>& By, const Vector3<T>& Bz) {
-    RotationMatrix<T> R(DoNotInitializeMemberFields{});
+    RotationMatrix<T> R;
     R.SetFromOrthonormalColumns(Bx, By, Bz);
     return R;
   }
@@ -192,7 +192,7 @@ class RotationMatrix {
   /// columns of `R_AB` are `Bx`, `By`, `Bz`.
   static RotationMatrix<T> MakeFromOrthonormalRows(
       const Vector3<T>& Ax, const Vector3<T>& Ay, const Vector3<T>& Az) {
-    RotationMatrix<T> R(DoNotInitializeMemberFields{});
+    RotationMatrix<T> R;
     R.SetFromOrthonormalRows(Ax, Ay, Az);
     return R;
   }
@@ -583,10 +583,6 @@ class RotationMatrix {
   // epsilon) used to check whether or not a rotation matrix is orthonormal.
   static constexpr double kInternalToleranceForOrthonormality_{
       128 * std::numeric_limits<double>::epsilon() };
-
-  // Constructs a RotationMatrix without initializing the underlying 3x3 matrix.
-  struct DoNotInitializeMemberFields{};
-  explicit RotationMatrix(DoNotInitializeMemberFields) : R_AB_() {}
 
   // Constructs a %RotationMatrix from a Matrix3.  No check is performed to test
   // whether or not the parameter R is a valid rotation matrix.

